Dev guide: review read data (#8666)
pingcap/docs: develop/dev-guide-use-stale-read.md

5 rows in set (0.02 sec)
```

In the list at this time (2022-04-20 15:20:00), the price of *The Story of Droolius Caesar* is 100.0.

> At the same time, the seller found that the book was very popular, so he raised the price of the book to 150.0 through the following SQL statement:

Reviewer: Do not use "he" when it could be "he" or "she".

5 rows in set (0.01 sec)
```

> If it is not necessary to use the latest data, you can query with Stale Read, which might return outdated data, to avoid the latency caused by data synchronization during a strongly consistent read.

Reviewer: ```suggestion
If it is not necessary to use the latest data, you can query with Stale Read, which might return outdated data, to avoid the latency caused by data replication during a strongly consistent read.
```

5 rows in set (0.01 sec)
```

In addition to specifying the exact time, you can also use:

> - `AS OF TIMESTAMP NOW() - INTERVAL 10 SECOND` instructs to query the latest data 10 seconds ago.
> - `AS OF TIMESTAMP TIDB_BOUNDED_STALENESS('2016-10-08 16:45:26', '2016-10-08 16:45:29')` instructs to query the latest data between `2016-10-08 16:45:26` and `2016-10-08 16:45:29`.
> - `AS OF TIMESTAMP TIDB_BOUNDED_STALENESS(NOW() -INTERVAL 20 SECOND, NOW())` instructs to query the latest data within 20 seconds.

Reviewer: ```suggestion
- `AS OF TIMESTAMP NOW() - INTERVAL 10 SECOND` queries the latest data 10 seconds ago.
- `AS OF TIMESTAMP TIDB_BOUNDED_STALENESS('2016-10-08 16:45:26', '2016-10-08 16:45:29')` queries the latest data between `2016-10-08 16:45:26` and `2016-10-08 16:45:29`.
- `AS OF TIMESTAMP TIDB_BOUNDED_STALENESS(NOW() -INTERVAL 20 SECOND, NOW())` queries the latest data within 20 seconds.
```

5 rows in set (0.01 sec)
```

> After committing the transaction with the `COMMIT;` statement, it is able to read the latest data.

Reviewer: ```suggestion
After the transaction with the `COMMIT;` statement is committed, you can read the latest data.
```

With the `SET TRANSACTION READ ONLY AS OF TIMESTAMP` statement, you can set the opened transaction or the next transaction to be a read-only transaction based on a specified historical time. The transaction will read historical data based on the provided historical time.

<SimpleTab>
<div label="SQL" href="next-txn-sql">

> The following statement is used to switch the opened transaction to read-only mode. With the `AS OF TIMESTAMP` clause, you can read historical data 5 seconds ago through the `AS OF TIMESTAMP` statement.

Reviewer: ```suggestion
For example, you can use the following `AS OF TIMESTAMP` statement to switch the ongoing transactions to the read-only mode and read historical data 5 seconds ago.
```

```sql
SET TRANSACTION READ ONLY AS OF TIMESTAMP NOW() - INTERVAL 5 SECOND;
```

</div>
<div label="Java" href="next-txn-java">

> Defining a helper class for transactions, which encapsulates the command to open a transaction level Stale Read as a helper method.

Reviewer: ```suggestion
You can define a helper class for transactions, which encapsulates the command to enable Stale Read at the transaction level as a helper method.
```

## Session level

> To support reading historical version data, TiDB has introduced a new system variable `tidb_read_staleness` since v5.4. It is used to set the range of historical data that the current session is allowed to read. Its data type is `int` and its scope is `SESSION`.

Reviewer: ```suggestion
To support reading historical data, TiDB has introduced a new system variable `tidb_read_staleness` since v5.4. you can use it to set the range of historical data that the current session is allowed to read. Its data type is `int` and its scope is `SESSION`.
```

<SimpleTab>
<div label="SQL">

Enable Stale Read in a session:

{{< copyable "sql" >}}

```sql
SET @@tidb_read_staleness="-5";
```

For example, if the value is set to -5, on the condition that TiKV has the corresponding historical version's data, TiDB selects a timestamp as new as possible within a 5-second time range.

> Close Stale Read in the session:

Reviewer: ```suggestion
Disable Stale Read in the session:
```
